pyop2/caching.py

```diff
@@ -440,119 +441,66 @@ class DEFAULT_CACHE(dict):
-                        cache_hit = False
-                    else:
-                        debug(debug_string + "hit")
-                        cache_hit = True
-                    all_present = comm.allgather(cache_hit)
-
-                    # If not present in the cache of all ranks we force re-evaluation on all ranks
-                    if not min(all_present):
-                        value = CACHE_MISS
-
-                if value is CACHE_MISS:
-                    value = func(*args, **kwargs)
-                return local_cache.setdefault(key, value)
-
-            return wrapper
-        return decorator
-else:
-    def parallel_cache(
-        hashkey=default_parallel_hashkey,
-        comm_fetcher=default_comm_fetcher,
-        cache_factory=lambda: DEFAULT_CACHE(),
-    ):
-        """Parallel cache decorator.
-        """
-        def decorator(func):
-            @PETSc.Log.EventDecorator("PyOP2 Cache Wrapper")
-            @wraps(func)
-            def wrapper(*args, **kwargs):
-                """ Extract the key and then try the memory cache before falling back
-                on calling the function and populating the cache.
-                """
-                k = hashkey(*args, **kwargs)
-                key = _as_hexdigest(*k), func.__qualname__
-                # Create a PyOP2 comm associated with the key, so it is decrefed when the wrapper exits
-                with temp_internal_comm(comm_fetcher(*args, **kwargs)) as comm:
-                    # Fetch the per-comm cache_collection or set it up if not present
-                    # A collection is required since different types of cache can be set up on the same comm
-                    cache_collection = comm.Get_attr(comm_cache_keyval)
-                    if cache_collection is None:
-                        cache_collection = {}
-                        comm.Set_attr(comm_cache_keyval, cache_collection)
-                    # If this kind of cache is already present on the
-                    # cache_collection, get it, otherwise create it
-                    local_cache = cache_collection.setdefault(
-                        (cf := cache_factory()).__class__.__name__,
-                        cf
-                    )
-                    local_cache = cache_collection[cf.__class__.__name__]
-
-                    # If this is a new cache or function add it to the list of known caches
-                    if (comm, comm.name, func, local_cache) not in [(c.comm, c.comm_name, c.func, c.cache()) for c in _KNOWN_CACHES]:
-                        # When a comm is freed we do not hold a reference to the cache.
-                        # We attach a finalizer that extracts the stats before the cache
-                        # is deleted.
-                        _KNOWN_CACHES.append(_CacheRecord(next(_CACHE_CIDX), comm, func, local_cache))
-
-                    value = local_cache.get(key, CACHE_MISS)
-
+def parallel_cache(
```

Comment:
Previously we duplicated this logic to avoid a `comm.allgather`. I have determined that this is *essential* to avoid deadlocks (hits on not all of the ranks will otherwise deadlock). Since both code paths are now the same apart from a handful of `debug()` statements I have removed the fork.

Reply:
I thought we were caching on `comm`. How can we get to the situation where we have cache_hit on some and cache_miss on the others?

Reply:
It happens when we disk cache things, which uses the same code path

Reply:
Though I do not actually understand how, apart from having race conditions, we could actually have a hit on some ranks but not others. Ideas welcome.

Reply:
So it turns out that the problem was actually that the *keys* were not consistent between ranks. I have enabled a check for this when `PYOP2_SPMD_STRICT` is enabled (which it now is in CI) and this has thrown up a number of times where we are doing the wrong thing and risking deadlocks.

Reply:
That makes sense. I think this is a lot better now. I'm not sure if I understand `PYOP2_SPMD_STRICT` variable. Do you know when we do not have SPMD and that non-SPMD is relevant in `parallel_cache`?

Reply:
`PYOP2_SPMD_STRICT` basically just adds a couple of checks in places to assert that things behave collectively as expected.

I don't know what you mean about "when we do not have SPMD".

Reply:
If it is just for checking, I think I'm fine with that.